internal/controller/preservation/helper.go

```diff
@@ -0,0 +1,305 @@
+	// - The ObjectMeta includes the original annotations and labels, along with an additional label indicating
+	//   preservation.
+	// - The original Secret type is preserved to maintain compatibility with future restores.
+	// - The Data field uses the original resource name as the key and stores the marshaled data as the value.
+	//   This ensures that the original Secret remains unaltered and can be restored later when required.
+	preservedSecret := &corev1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:        preservedName,
+			Namespace:   namespace,
+			Annotations: originalSecret.Annotations,
+			Labels:      addPreservedLabel(originalSecret.Labels),
+		},
+		Type: originalSecret.Type,
+		Data: map[string][]byte{
+			name: encodedData,
+		},
+	}
+
+	// Create the backup Secret.
+	if err := c.Create(ctx, preservedSecret); err != nil {
+		return fmt.Errorf("failed to create preserved Secret %s/%s: %w", namespace, preservedName, err)
+	}
+
+	return nil
+}
+
+// restoreConfigMap restores a preserved ConfigMap to its original state.
+// It retrieves the preserved ConfigMap by name, decodes its data, and creates a new ConfigMap
+// with the original name and metadata.
+func restoreConfigMap(
+	ctx context.Context,
+	c client.Client,
+	namespace, preservedName string,
+) error {
+	// Fetch the preserved ConfigMap from the cluster.
+	preservedConfigMap := &corev1.ConfigMap{}
+	key := types.NamespacedName{Namespace: namespace, Name: preservedName}
+	if err := c.Get(ctx, key, preservedConfigMap); err != nil {
+		return fmt.Errorf("failed to retrieve ConfigMap %s/%s for restoration: %w", namespace, preservedName, err)
+	}
+
+	// Validate that the preserved ConfigMap contains exactly one key.
+	if len(preservedConfigMap.Data) != 1 {
+		return fmt.Errorf("preserved ConfigMap should only have 1 key, received %d", len(preservedConfigMap.Data))
+	}
+
+	// Extract and decode the data from the preserved ConfigMap.
+	originalName := maps.Keys(preservedConfigMap.Data)[0]
+	var decodedData map[string]string
+	if err := json.Unmarshal([]byte(preservedConfigMap.Data[originalName]), &decodedData); err != nil {
+		return fmt.Errorf("failed to unmarshal ConfigMap.Data, err: %w", err)
+	}
+
+	// Construct the restored ConfigMap using the original resource name and metadata.
+	// - The name is set to the original resource's name.
+	// - The namespace, annotations, and labels are copied from the preserved ConfigMap,
+	//   with the preservation label removed to avoid marking it as a backup.
+	// - The data is decoded from the preserved ConfigMap, restoring the original structure.
+	restoredConfigMap := &corev1.ConfigMap{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:        originalName,
+			Namespace:   namespace,
+			Annotations: preservedConfigMap.Annotations,
+			Labels:      removePreservedLabel(preservedConfigMap.Labels),
+		},
+		Data: decodedData,
+	}
+
+	// Attempt to create the restored ConfigMap. Do not error if the ConfigMap already exists.
```

Comment:
Why? Would this not be an error we want to care about?

If it already exists should we patch it with the backed up data?